pkg/cmd/rollbackcopy/etcdclientutils.go

```
@@ -0,0 +1,137 @@
+package rollbackcopy
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"google.golang.org/grpc"
+	"io"
+	"k8s.io/klog"
+	"os"
+	"time"
+
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/pkg/transport"
+)
+
+func getEtcdClient(endpoints []string) (*clientv3.Client, error) {
+	dialOptions := []grpc.DialOption{
+		grpc.WithBlock(), // block until the underlying connection is up
+	}
+
+	tlsInfo := transport.TLSInfo{
+		CertFile:      os.Getenv("ETCDCTL_CERT"),
+		KeyFile:       os.Getenv("ETCDCTL_KEY"),
+		TrustedCAFile: os.Getenv("ETCDCTL_CACERT"),
+	}
+	tlsConfig, err := tlsInfo.ClientConfig()
+
+	cfg := &clientv3.Config{
+		DialOptions: dialOptions,
+		Endpoints:   endpoints,
+		DialTimeout: 2 * time.Second,
+		TLS:         tlsConfig,
+	}
+
+	cli, err := clientv3.New(*cfg)
+	if err != nil {
+		return nil, fmt.Errorf("failed to make etcd client for endpoints %v: %w", endpoints, err)
+	}
+	return cli, nil
+}
+
+func getClusterVersionAndUpgradeInfo(cli *clientv3.Client) (string, bool, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	clusterversionKey := "/kubernetes.io/config.openshift.io/clusterversions/version"
+	resp, err := cli.Get(ctx, clusterversionKey)
+	if err != nil {
+		return "", false, err
+	}
+	klog.Info("### Response for client GET: ", resp)
+	if len(resp.Kvs) != 1 {
+		klog.Errorf("Expected to get a single key from etcd, got %d", len(resp.Kvs))
+		return "", false, fmt.Errorf("Expected to get a single key from etcd, got %d", len(resp.Kvs))
```

Wrap and return error instead of logging here